### ISSUE TITLE
WDA 4.7.0 and below did not return `\n` but 4.8.0+ returned `\n` in a label

### ISSUE THREAD

**Issue by KazuCocoa** [appium/appium#17554]
## The problem

[updated]
A new line in XML is expected.

[`<element id>/text` endpoint](https://github.com/appium/WebDriverAgent/blob/a1824895274c60798b506f31054e4ce1e0400c71/WebDriverAgentLib/Commands/FBElementCommands.m#L133) returned `line break\nline break` with `\n` result since https://github.com/appium/WebDriverAgent/pull/596, while the same one returned `line break line break` without `\n` before the change.

`@driver.find_element :xpath, "//XCUIElementTypeStaticText[@name='line break line break']"` responded the element in both version, WDA 4.7.0 (before the PR) and 4.8.0 (with the PR). `@driver.find_element :xpath, "//XCUIElementTypeStaticText[@name='line break\nline break']"` did not even 4.8.0 since the XML source did not have `\n` as below.

The `description` format by vanilla XCTest had `\n` like `label: 'line break\nline break'` in both versions, so potentially https://github.com/appium/WebDriverAgent/pull/596 fixed this behavior(?), but in the case, our XML source did not include `\n`, so find an element 

The below two lines `line break` is this issue. 

![Screen Shot 2022-10-03 at 23 50 06](https://user-images.githubusercontent.com/5511591/193752913-29f5c2c8-6528-4608-86f4-1a5cdb919157.png)

The page source in JSON is below with WDA 4.7.0 and 4.8.0.
```
                                      {
                                        "isEnabled" : "1",
                                        "isVisible" : "1",
                                        "isAccessible" : "1",
                                        "frame" : "{{76, 402}, {238, 32}}",
                                        "rect" : {
                                          "y" : 402,
                                          "x" : 76,
                                          "width" : 238,
                                          "height" : 32
                                        },
                                        "value" : "line break line break",
                                        "label" : "line break line break",
                                        "type" : "StaticText",
                                        "name" : "line break line break",
                                        "rawIdentifier" : null
                                      }
                                    ],
```

NSPredicate like `@driver.find_element :predicate,"label == 'line break\\nline break'"` found the element in both 4.7.0 and 4.8.0, thus as iOS, with `\n` has been correct way, I assume. (`@driver.find_element :predicate,"label == 'line break line break'"` did not find the element.)

So... a text with a new line behavior changed since 4.8.0, and kind of mixed after the PR in the page source and XML stuff. I have't still figure out the fix yet, but maybe it would be nice to follow with `\n` as a breaking change in the WDA to make the behavior as same as vanilla XCTest. Then, we should update the XML page source to include it properly(?).

In my investigation so far, the difference comes from `self.label` in https://github.com/appium/WebDriverAgent/blob/master/WebDriverAgentLib/Categories/XCUIElement%2BFBWebDriverAttributes.m#L83 .
4.7.0 returned `line break line break` (without `\n`) while 4.8.0 returned `line break\nline break`.
`po` result is:

4.7.0
```
(lldb) po label
line break line break
```

4.8.0
```
(lldb) po label
line break
line break
```

## Environment

* Appium version (or git revision) that exhibits the issue: WDA 4.8.0
* Last Appium version that did not exhibit the issue (if applicable): WDA 4.7.0
* Desktop OS/version used to run Appium:
* Node.js version (unless using Appium.app|exe): nodejs 16
* Npm or Yarn package manager: npm
* Mobile platform/version under test: ios 15.5 in my env, but this may happen other versions as well. Xcode was 13.4.
* Real device or emulator/simulator: both
* Appium CLI or Appium.app|exe:

## Code To Reproduce Issue [ Good To Have ]

You can find this new lines text in [iOSWebView.app.zip](https://github.com/appium/appium/files/9703618/iOSWebView.app.zip)
 (a sample app for simulator).
SPECIAL CHARACTERS -> `\n LINE BREAK` -> then, the alert prepare

**Comment by KazuCocoa** [comment]
cc @mykola-mokhnach

**Comment by mykola-mokhnach** [comment]
Just to clarify it - XCUIElement attributes, like label or value do contain the line break, but the actual XML page source ignores it. The expected result is that the line break must be properly escaped in all XML attributes rather than ignored, right?

**Comment by mykola-mokhnach** [comment]
I've also checked the current xml transformation logic. Basically, a label like `a\nb` gets translated to the following XML attribute value: `a&#10;b`

**Comment by KazuCocoa** [comment]
hm, ok, then, 4.7.0 did not include a new line in the label (I mean via `/element/:uuid/text`'s result.), but 4.8.0 included a new line now as same as vanilla XCUIElement attributes.

So, like the below assertion was passed in 4.7.0 and below, but not in 4.8.0 since `el.text` included `\n`.

```
el = @driver.find_element :xpath, "//XCUIElementTypeStaticText[@name='line break line break']"
assert el.text == 'line break line break'
```

> the actual XML page source ignores it
oh, i see. I didn't remember this point then. Then, this issue is the result of `.text`'s new line handling, i think

**Comment by mykola-mokhnach** [comment]
> hm, ok, then, 4.7.0 did not include a new line in the label (I mean via /element/:uuid/text's result.), but 4.8.0 included a new line now as same as vanilla XCUIElement attributes.

I'd say this is correct

el = @driver.find_element :xpath, "//XCUIElementTypeStaticText[@name='line break line break']"

I assume the updated query would look like `"//XCUIElementTypeStaticText[@name='line break&#10;line break']"`
The question is whether we'd like to represent this 0xA and 0xD chars differently, or keep it as is or maybe strip them completely, or make this configurabel

**Comment by mykola-mokhnach** [comment]
My proposal would be to replace 0x09 with `\t`, 0x0A with `\n` and 0x0D with `\r` in the XML (and this is xpath queries) to avoid unnecessary confusions

**Comment by KazuCocoa** [comment]
The XML query itself worked in both 4.7.0 and 4.8.0 with `//XCUIElementTypeStaticText[@name='line break line break']` right now, so the difference was the attribute result via `handleGetText` in WDA, 4.7.0 did not have `\n` but 4.8.0 had.

So, as WDA, 4.8.0 behaves more close to vanilla XCUIElement attribute.
Then, probably we need to address this diff in https://github.com/appium/appium-xcuitest-driver#element-attributes this new line diff. let me create a pr for it.

The original report was by a friend of mine, and combination of XML and the attribute result change. So... for now, you can ignore if WDA needs to handle a new line with escaping it right now.

(sorry, the page source stuff was confused in the description..)

**Comment by Dan-Maor** [comment]
I found the cause of the different behavior.
In `fb_filterDescendantsWithSnapshots` the final query in 4.8.0 is generated with a block rather than a predicate string, which causes a snapshot with different values to be returned.
Before:
> query = [query matchingPredicate:[NSPredicate predicateWithFormat:@"%K IN %@", FBStringify(XCUIElement, wdUID), sortedIds]];

After:
> NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(XCElementSnapshot *snapshot, NSDictionary *bindings) {
    return [sortedIds containsObject:(NSString *)snapshot.fb_uid];
  }];
  query = [query matchingPredicate:predicate];

I checked out WDA 4.7.0 and replaced the predicate with the new block predicate (used Xcode 13.2), and the returned string of the `/text` endpoint was the same as 4.8.0.

Should we dig deeper to see if we can preserve the previous behavior? Or - if the result is more in line with Vanilla XCTest - should we keep the change as is?
